cmds/tail/tail.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Tail prints the lasts 10 lines of a file. Can additionally follow the
+// the end of the file as it grows.
+//
+// Synopsis:
+//     tail [-f] [-n lines_to_show] [FILE]
+//
+// Description:
+//     If no files are specified, read from stdin.
+//
+// Options:
+//     -f: follow the end of the file as it grows
+//     -n: specify the number of lines to show (default: 10)
+
+// Missing features:
+// - follow-mode (i.e. tail -f)
+
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"syscall"
+)
+
+var (
+	flagFollow   = flag.Bool("f", false, "follow the end of the file")
+	flagNumLines = flag.Int("n", 10, "specify the number of lines to show")
+)
+
+// getBlockSize returns the block size for the disk where the file pointed by
+// the reader resides
+func getBlockSize(f os.File) (int64, error) {
+	var stat syscall.Stat_t
+	err := syscall.Stat(f.Name(), &stat)
+	if err != nil {
+		return 0, err
+	}
+	return stat.Blksize, nil
+}
+
+// readLastLinesBackwards reads the last N lines from the provided file, reading
+// backwards from the end of the file. This is more efficient than reading from
+// the beginning, but can only be done on seekable files, (e.g. this won't work
+// on stdin). For non-seekable files see readLastLinesFromBeginning.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesBackwards(inFile os.File, writer io.Writer, numLines int) error {
+	if numLines < 0 {
+		return fmt.Errorf("The number of lines cannot be negative")
+	}
+	// get the file's block size. This will be the size of each chunk to read
+	blkSize, err := getBlockSize(inFile)
+	if err != nil {
+		return err
+	}
+	// go to the end of the file
+	lastPos, err := inFile.Seek(0, os.SEEK_END)
+	if err != nil {
+		return err
+	}
+	// read block by block backwards until `numLines` lines are found
```

Comment:
why all this seeking? Why not just pread? Then it gets pretty easy. First off, most lines are short. A good heuristic is to take the # lines you want, multiply by 81, then do a pread for the last 81*n bytes of the file, split on \n, and see if you have enough lines. I think the blocksize test is way more than you need to do.

Also, for fun, write a simple Go program that traverses our entire disk, and get a histogram of line lengths -- I'll bet it's under 16 chars per line ...

Reply:
very good point, I didn't consider `pread` (maybe just because I rarely use it). Will rewrite using it, thanks!

Reply:
sorry about that, I meant ReadAt.